### DISCUSSION TITLE
[Sieghard branch] New Dialogs components, fixes for msestatfile.pas and db files from Sieghard.

### DISCUSSION THREAD

**Discussion by fredvs** [mse-org/mseide-msegui#87] [General]
Hello.

Sieghard has done an amazing job adding new features to the Dialogs components and fixes for the stat and db files.

I tried the **newdialogs** components on Linux and everything seems ok, same for the stat files.
For db files, because I don't know how thoroughly to test it, it's on the to-do list.

It would be great if other people tested all this code on Windows as well.

Here are Sieghard's compressed files.

**newdialog.zip**, paste-replace all the files in **/mseide-msegui/lib/common/dialogs/**
[newdialogs.zip](https://github.com/mse-org/mseide-msegui/files/15082697/newdialogs.zip)

**msegui_memorystatfile_patches.zip**, paste-replace msestatfile.pas in **/mseide-msegui/lib/common/kernel/**
[msegui_memorystatfile_patches.zip](https://github.com/mse-org/mseide-msegui/files/15082713/msegui_memorystatfile_patches.zip)

**msegui_lib_db_patches.zip**, paste-replace all the files in **/mseide-msegui/lib/common/db/**
[msegui_lib_db_patches.zip](https://github.com/mse-org/mseide-msegui/files/15082724/msegui_lib_db_patches.zip)

---------------------------------------------------------------------------------

There is also new sql browsers:
[SQLiteBrowse.zip](https://github.com/mse-org/mseide-msegui/files/15082819/SQLiteBrowse.zip)

[PSQLbrowse.zip](https://github.com/mse-org/mseide-msegui/files/15082876/PSQLbrowse.zip)

Data-demo for testing:
[PV_Data.zip](https://github.com/mse-org/mseide-msegui/files/15082856/PV_Data.zip)

-------------------------------------------------------------------------

And the famous mse-clock:
[MSEclock.zip](https://github.com/mse-org/mseide-msegui/files/15083101/MSEclock.zip)

---------------------------------------------

**Comment by rchastain** [comment]
Hello Fred!

Which features? Fixes for what? Sorry, but I have no time for _pasting-replacing_ anything without knowing what I do.

Why Sieghard doesn't create a Github account and fork the repository? It isn't so complicated.

**Reply by fredvs** [reply]
Hello Roland!

I think that Sieghard want to preserve his privacy.

There is a readme.txt file in **newdialogs.zip**:

-----------------------------------------------------

New features for "newdialogs”                     So 7. Apr 13:21:45 CEST 2024

Unit "vectors”
  Completely new set of arithmetic operators for 2-dimensional vectors of "PointTy"
  OPERATOR  +  adds £wo vectors, i.e. appends one to the other
  OPERATOR  -  subtracts two vectors, i.e. appands the negative of the second to the first one
  OPERATOR  =  compares two vectors, returns "true" if both components match
  OPERATOR <>  compares two vectors, returns "true” if at least one component DOES NOT match
  OPERATOR  *  calculates the scalar product of a vector
  OPERATOR  *  multiplies a vector with an integer and returns the resulting larger vector
               (sequence if operands does not matter)
  OPERATOR DIV divides a vector by an integer and returns the resulting reduced vector
  OPERATOR := assigns an array of two integers to the components of a vector


Unit "msedialog":

  New type definition:

    dialogposty - enumeration type to specify dialog window positioning
      dp_none:                 no positioning, use form setting
      dp_mousepos:             show dialog centered at current mouse position
      dp_defaultpos:           show dialog at design position or position set after form creation
      dp_screencentered:       show dialog centered on screen
      dp_screencenteredvirt:   show dialog centered on virtual screen
      dp_transientforcentered: show dialog centered on currently active window (?)
      dp_mainwindowcentered:   show dialog centered on application main window

  New functionality for "tdialogform":
    Two new constructors adding positioning and state memory

      CONSTRUCTOR Create (Sender: TComponent; where: dialogposty = dp_none);
        Allows specification of dialog positioning according to parameter "where".
        If "where" isn't specified, uses form setting

      CONSTRUCTOR Create (Sender: TComponent; CONST StatName: msestring; where: dialogposty);
        Allows specification of dialog positioning according to parameter "where" as above,
        additionaly specifies memory statfile name for keeping state memory between invocations.

  New ancillary functions:

    FUNCTION keepOnScreen (CONST Sender: twidget; shift: PointTy): PointTy;
      Calculates new position for "Sender" window to put it at a position "shift" picels away
      from its current one, new position should be assigned to "Window.decoratedPos" to keep
      window frame on screen also.

    PROCEDURE registerSavedDialogs (StateFile: tStatFile; CONST DialogNames: msestringarty);
    PROCEDURE registerSavedDialogs (BaseForm: tmseform; CONST DialogNames: msestringarty);
      These procedures allow saving dialog state memory over application invocations for the
      (accordingly created) dialogs whose "StatName"s are specified by "DialogNames".
      The first version keeps these data through a directly specified sz´tatfile "StateFile",
      while the second uses the statfile set up for the form "BaseForm", usually the
      application main form.
      --- REMARK:
          With the modified stat file processing as of 2024-04, the dialog states can be
          perpetuated "automatically" by defining an "onstatupdate" procedure like this:

      PROCEDURE <Form class name>.AssessStates (CONST sender: TObject; CONST filer: tstatfiler);
       VAR
         MemFiles: msestringarty;
       BEGIN
         IF filer IS tStatReader THEN BEGIN
           MemFiles:= [''];
           MemFiles:= (filer AS tStatReader).ReadArray ('savedmemoryfiles', MemFiles);
           registerSavedDialogs (PStateFile, MemFiles);
         END;
       //  IF filer IS tStatWriter THEN BEGIN END; -- not needed here!
       END;


Unit "msecolordialog":

  New functionality for calling function "colordialog":
    This function now also takes a (final) parameter "providedform" of type "tcolordialogfo"
    that can be preinitialized to application specific properties.
    If no such form is passed, it is MODIFIED in such a manner as to NO LONGER show up FIXED
    at its design position, but to appear at the current mouse position.
    This ALTERS the behaviour if used from within a "tcoloredit" and possibly other ones too.


Unit "msesearchreplace":
  New calling interface and functionality for class "treplacedialogfo":
    CONSTRUCTOR Create (CONST Sender: TComponent; replacer: boolean = true;
                        where: dialogposty = dp_none); OVERLOAD;
    CONSTRUCTOR Create (CONST Sender: TComponent; CONST StatName: msestring;
                        replacer: boolean = true; where: dialogposty = dp_none); OVERLOAD;
      The constructor now already defines the kind of dialog created and the positioning method.
      Positioning is handled like for a simple "tdialogform" and preset to "dp_none" as well.
      Parameter "replacer" specifies the kind of dislog created:
        the default setting of "true" yields a replace dialog,
        the explicit value "false" yields a search-only dialog.

------------------------------------

**Comment by rchastain** [comment]
Or else (if Sieghard doesn't want to create a Github account) you could create a new branch and push on that branch the modifications.

**Reply by fredvs** [reply]
Yes, but for that I have to remember how to do this... :-)

**Reply by rchastain** [reply]
Sincerely, it's very simple. Or if you don't feel able to do it, just fork the repository and push the modifications on the new repository. So that we can see the modifications... I can even do it if you want.

**Comment by fredvs** [comment]
For the SQL demos, maybe they could be added in the Sieghard directory of MSEUniverse:
https://github.com/mse-org/mseuniverse/tree/main/attic/msedocumenting/mse/trunk/contributed/sieghard

By the way, I would vote for a cleaning-reorganisation of all the demos-directories of MSEUniverse.

**Reply by rchastain** [reply]
Yes, good idea.

**Reply by rchastain** [reply]
Took the liberty to upload the clock [here](https://codeberg.org/rchastain/mseuniverse/src/branch/master/attic/contributed/sieghard).

The path is too long and complicated. So, I approve the idea of a reorganisation.

[edit]
Reorganized my directory and replaced the previous link.
[/edit]

**Reply by rchastain** [reply]
What means **_attic_**?

**Reply by rchastain** [reply]
Suggestion of a simple reorganization of the _attic_ folder:

https://codeberg.org/rchastain/mseuniverse/src/branch/master/attic

I just deleted superfluous folders and move their content to root (to _attic_).

**Reply by fredvs** [reply]
>Took the liberty to upload the clock [here](https://codeberg.org/rchastain/mseuniverse/src/branch/master/attic/msedocumenting/mse/trunk/contributed/sieghard/clock).

Nice, many thanks.

>What means attic?

https://en.wikipedia.org/wiki/Attic

(But I still dont understand the meaning...)

**Reply by fredvs** [reply]
>Suggestion of a simple reorganization of the attic folder:

Ho, yes, much better.
Please commit-push it to https://github.com/mse-org/mseuniverse

Thanks for this great and indispensable work.

**Comment by fredvs** [comment]
Hello.

There is a new branch "sieghard" with all the files: https://github.com/mse-org/mseide-msegui/tree/sieghard

**Comment by rchastain** [comment]
Good job Fred.

I installed it and built it under Linux (with the script attached below).

I cannot configure MSEide: I cannot select a value for MSEDIR. When I try to select a directory, the Settings window closes itself and the value is not saved.

Maybe I did something wrong...

[installmseide-sieghard.zip](https://github.com/mse-org/mseide-msegui/files/15179599/installmseide-sieghard.zip)

**Comment by fredvs** [comment]
Hello Roland.

I ran your script (woow out-of-the-box).
Indeed, when choosing a directory, after selection, the Settings window closes and nothing is saved.

;-(

**Comment by fredvs** [comment]
The good news is that other dialogs are working.
So, added manually the directory of MSEDIR (not via the dialog button) is ok.
The **Tfilenameedit** dialog component needs checkup.

One of the very good things, with MSEide using Sieghard's newdialogs, all the dialogs are in the correct position, much better than the original dialogs which are in a random position.
You can try/compare by clicking on the open file or in the project menu.

**Comment by fredvs** [comment]
Hello Roland!

Your install script is very good.
I took inspiration for a script to install ideU desktop icon, many thanks.

Note that I did set:
`StartupNotify=false `

Otherwise there is a Hour-Glass icon for the cursor when I go outside the application.

**Comment by fredvs** [comment]
There is a new branch "sieghard" with last fixes: https://github.com/mse-org/mseide-msegui/issues/89#issuecomment-2093929190

Tests are welcome!

**Comment by rchastain** [comment]
Indeed the issue of configuration window is solved. Good job!

I will take time to make some more tests. Are the MSEClock and the SQL browsers included somewhere, or do we have to use the ZIP files attached above?

**Reply by fredvs** [reply]
The zip files of MSEclock and SQL browser + data are here: https://github.com/mse-org/mseide-msegui/discussions/87#discussion-6550805

**Comment by rchastain** [comment]
> I think that Sieghard want to preserve his privacy.

Sieghard's work appears to be of very good quality. It would be so simpler, for him and for us, if Sieghard would accept to create a GitHub account. He could use a nickname. He could even (despite I see no reason for that) make a private fork and make it visible only for us.

Git can look complicated at first sight but in fact three or four commands are enough to do all the job. And it's much more efficient than playing with ZIP files and emails. The MSE project uses git for a long time!

**Reply by fredvs** [reply]
>Sieghard's work appears to be of very good quality

Sieghard's work is of highest quality and so precious.

>if Sieghard would accept to create a GitHub account

Yes indeed and of course it would be great if Sieghard had a git account.
Note that there is also a mirror site on CodeBerg:
https://codeberg.org/fredvs/mseide-msegui

And another mirror site on GitLab:
https://gitlab.com/fredvs/mseide-msegui

**Comment by rchastain** [comment]
> Ho, yes, much better. Please commit-push it to https://github.com/mse-org/mseuniverse

Done.

**Comment by fredvs** [comment]
Hello.

Sieghard gave new fixes for the new dialogues.
This was committed in the Sieghard branch just now.

There is also a new version of MSEclock and FileRequester a new demo.
With a readme.txt in each zip file.

[MSEclock_240509.zip](https://github.com/mse-org/mseide-msegui/files/15255684/MSEclock_240509.zip)

[FileRequest.zip](https://github.com/mse-org/mseide-msegui/files/15255686/FileRequest.zip)

Enjoy.

**Reply by rchastain** [reply]
Thanks Fred. I will take a look.

**Comment by fredvs** [comment]
About **FileRequest** project, I did not catch how to use it.
It needs to be run via a terminal and using some parameters but I dont get how to use it.
A example would be welcome (readme.txt did not help me).

**Comment by rchastain** [comment]
I don't know what is new in the new version of MSEClock, but the bug that I reported is still here.

**Reply by fredvs** [reply]
Moving the window clicking on the form maintaining the button and moving the mouse is ok now.
But yes, the resize-bug is still there, see here: https://github.com/mse-org/mseide-msegui/issues/90#issuecomment-2101639864

**Comment by rchastain** [comment]
My attempt to build **FileRequest** project:

```
Warning: You are using the obsolete switch -OG
Warning: You are using the obsolete switch -Op<nr>, please use -Op<name>
Note: DWARF debug information cannot be used with smart linking on this target, switching to static linking
Free Pascal Compiler version 3.2.2 [2021/05/16] for x86_64
Copyright (c) 1993-2021 by Florian Klaempfl and others
Target OS: Linux for x86-64
Compiling ./FileRequest.pas
Compiling /home/roland/Applications/mseide-sieghard/lib/common/kernel/msegui.pas
Compiling /home/roland/Applications/mseide-sieghard/lib/common/fpccompatibility/mclasses.pas
mclasses.pas(1205,46) Fatal: Can't find unit msesysintf used by mclasses
Fatal: Compilation aborted
```

**Reply by fredvs** [reply]
>mclasses.pas(1205,46) Fatal: Can't find unit msesysintf used by mclasses

you need to change in project option, Make / Directories (change TARGETDIR with TARGETOSDIR):

`  ${MSELIBDIR}kernel/$TARGETOSDIR`

or use

`  ${MSELIBDIR}kernel/linux`

(Not very out-of-the-box all this.)

**Reply by rchastain** [reply]
Thanks. I should have seen it.

**Comment by fredvs** [comment]
Hello.

Last fixes for newdialogs were committed in sieghard-branch: https://github.com/mse-org/mseide-msegui/commit/95e0554d98f4123098bf42ecc373c2afa54ed747

Here fixes of previous demos:

[SQLiteBrowse_240509.zip](https://github.com/mse-org/mseide-msegui/files/15264346/SQLiteBrowse_240509.zip)

[PSQLbrowse_240509.zip](https://github.com/mse-org/mseide-msegui/files/15264348/PSQLbrowse_240509.zip)

[FileRequest_240509.zip](https://github.com/mse-org/mseide-msegui/files/15264351/FileRequest_240509.zip)

**Comment by rchastain** [comment]
I could compile FileRequest. I did read the README. It is not very clear to me how to use it and what we can do with it. As you said, an example would have been welcome.

**Comment by fredvs** [comment]
Hello.
There are new fixes for /db files in sieghard branch.

**Comment by fredvs** [comment]
Hello.
There are new fixes for filedialogx in sieghard branch.
Also a demo for it:
[TextView.zip](https://github.com/user-attachments/files/17300527/TextView.zip)
